Kaleido Menu for Miscellaneous Collectibles

leggettc18 opened the pull request:
Adds a Kaleido menu for Miscellaneous collectibles, namely the ones that can be collected in rando but are not tracked in any existing Kaleido Menus. It can currently be toggled by pressing C-Up on the Quest Kaleido Page. After that it can be scrolled up and down via the control stick or D-Pad

There's probably some cleanup work left on this to go. For instance you currently can't move the cursor back to the opposite page turner (though you can still press the R or Z buttons to turn the page) and that's probably not a difficult thing to fix I just haven't yet. It's functional for its intended purpose as is and other than some navigation stuff it's ready for review. I will likely iterate more on this later, namely to add Dungeon Items (keys, maps, compasses, etc.), but I think what's here is enough for v1.

<!--- section:artifacts:start -->
### Build Artifacts
  - [soh.otr.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/2021130945.zip)
  - [soh-linux.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/2021136194.zip)
  - [soh-mac.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/2021139049.zip)
  - [soh-windows.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/2021139773.zip)
<!--- section:artifacts:end -->

leggettc18 committed:
Adds in-game display of certain rando collectibles.

Before, these were only available through the Item Tracker ImGui Window. With this commit, they can be accessed via holding C-Up on the Map Screen. Currently I've added Greg and Triforce Pieces (when applicable) to this menu. Boss Souls, Ocarina Buttons, and eventually Silver Rupees will be added later.

leggettc18 committed:
Adds ocarina buttons to in-game display

leggettc18 committed:
Initial pass on rendering the text on a black background.

leggettc18 committed:
Starting to render boss soul icons

leggettc18 committed:
Better alignment and rendering of Boss Soul icons.

leggettc18 committed:
Adds icons prefixing the list entries.

leggettc18 committed:
Switches boss souls to 32x32 icon.

leggettc18 committed:
Partially working Matrix/Vtx implementation

Currently hardcoded Greg text, replacing map screen completely.

leggettc18 committed:
Now rendering properly thanks to Archez!

leggettc18 committed:
Better implementation of accessing the new page.

- now attached to Quest status instead of Map
- now triggered by a toggle instead of holding a button
- now has its own background (temporarily save screen but will be replaced with something custom later)

leggettc18 committed:
Make KaleidoEntry's reactive to game state

Adds Greg proper and Triforce Hunt to the Misc. Collectibles Page.

leggettc18 committed:
Conditionally render Triforce Hunt

leggettc18 committed:
Documentation/Cleanup

leggettc18 committed:
WIP Ocarina Buttons rendering

leggettc18 committed:
Working ocarina buttons display

leggettc18 committed:
Renders buttons as Gray instead of using Grayscale

This may seem inconsistent, but with Grayscale they technically render as different shades of gray, especially with custom cosmetics. With this they now render as the same shade of gray.

leggettc18 committed:
Makes Ocarina Icon gray when no buttons have been collected.

leggettc18 committed:
Adds Boss Souls.

Currently they run off the menu, need to implement scrolling.

leggettc18 committed:
Implement Scrolling for the menu.

Need to figure out how to throttle the stick inputs still.

leggettc18 committed:
Moves input handling to draw function.

I hate it but that's how Kaleido does it and there's some input throttling logic in there, so in order to make this feel like a kaleido menu I have to also handle input in the draw function.

leggettc18 committed:
Removes custom cosmetic handling of Ocarina Buttons.

I've chosen not to respect the cosmetics for the sake of accessibility
and color contrast, but the code is still present and commented out in
case we want to reverse that decision.

leggettc18 committed:
Hopefully fixes mac build errors.

leggettc18 committed:
Implements update function via Hook.

leggettc18 committed:
Another mac fix hopefully

leggettc18 committed:
Cleans up unused code from the rectangle based attempt.

leggettc18 committed:
Clean up more unused code

inspectredc commented on soh/soh/Enhancements/kaleido.cpp:
```diff
@@ -0,0 +1,443 @@
+        }
+        if (ctx->GetOption(RSK_SHUFFLE_BOSS_SOULS).IsNot(RO_BOSS_SOULS_OFF)) {
+            static const char* bossSoulNames[] = {
+                    "Ghoma's Soul",
```
```suggestion
                    "Gohma's Soul",
```

inspectredc commented on soh/soh/Enhancements/kaleido.cpp:
```diff
@@ -0,0 +1,443 @@
+        }
+
+        Matrix_Translate(mX, mY, 0.0f, MTXMODE_APPLY);
+//        Matrix_Scale(0.75f, 0.75f, 0.75f, MTXMODE_APPLY);
```
```suggestion

```

inspectredc commented on soh/src/overlays/misc/ovl_kaleido_scope/z_kaleido_scope_PAL.c:
```diff
@@ -3920,6 +3939,11 @@ void KaleidoScope_Update(PlayState* play)
+                        Audio_PlaySoundGeneral(NA_SE_SY_DECIDE, &D_801333D4, 4, &D_801333E0, &D_801333E0, &D_801333E8);
+                        // this starts at 0, subtracting one and taking absolute value makes it 1, then subtracting
+                        // one again puts it back at 0. So this acts as a toggle.
+                        pauseCtx->randoQuestMode = ABS(pauseCtx->randoQuestMode - 1);
```
could this maybe be ^= 1 instead?

inspectredc reviewed (approved):
Looks good! Just one from from a functionality side too:

Currently toggling the rando kaleido menu will leave the cursor outline on the screen. I think ideally in the future the cursor gets moved to the right or left side, and free movement from the left to right sides (or vice versa) is possible like u mention in your PR. For now though think the next best thing would be to hide the cursor outline altogether.

Not for this PR but I'm still trying to figure out if there's a good way to visually indicate that this option exists (unless I missed it here), curious if you had any ideas

Pepper0ni commented:
I'll review this once the conflicts are fixed so i can test it on latest.

Malkierian commented:
Are Greg and the ocarina buttons the only things worth putting on there at the moment? The fishing pole has been added now, but I guess that's the only other thing I could think of that wasn't still somehow represented in the regular menu.

leggettc18 committed:
Commit Boss Soul icon

Pepper0ni commented:
Can confirm this is no longer crashing. Given recent glitchyness I will wait for that to be resolved before running a seed with this to properly test things

Pepe20129 commented on soh/soh/Enhancements/kaleido.cpp:
```diff
@@ -0,0 +1,443 @@
+    }
+
+    void KaleidoEntryOcarinaButtons::CalculateColors() {
+        // The commented out code in this function is for handling custom cosmetics for the
+        // A and C buttons. For contrast/accessibility reasons, I've chosen not to respect the
+        // custom cosmetics for now. I've left the code commented out in case that decision gets
+        // reversed later.
```
I'd argue if someone needs the cosmetics for accessibility reasons it should be accounted for.

Malkierian replied:
I would also like to see them account for cosmetics for accessibility purposes.

Pepe20129 commented on soh/soh/Enhancements/kaleido.cpp:
```diff
@@ -0,0 +1,443 @@
+    Kaleido::Kaleido() {
+        const auto ctx = Rando::Context::GetInstance();
+        int yOffset = 2;
+        mEntries.push_back(std::make_shared<KaleidoEntryIconFlag>(gRupeeCounterIconTex, G_IM_FMT_IA, G_IM_SIZ_8b, 16, 16,
+                              Color_RGBA8{ 0xC8, 0xFF, 0x64, 255 }, FlagType::FLAG_RANDOMIZER_INF,
+                              static_cast<int>(RAND_INF_GREG_FOUND), 0, yOffset, "Greg"));
```
Greg doesn't account for the green rupee color cosmetic.

leggettc18 replied:
I thought I remembered us deciding at some point that it shouldn't.

Malkierian replied:
Yeah, I think Greg at least should remain green. It's kind of intrinsic to the character of the item.

Pepe20129 reviewed (commented):
Looks good overall, I don't think inheritance is the best way of doing this but that's subjective, just a couple of small issues:
- It's not possible to move from the `L` button to the `R` button with the control stick when the menu is active.
- Some cosmetics related issues.

Pepper0ni commented:
I thought of something else random that could be tracked: whether or not you've freed epona. Not much use now, but eponasanity would be a potential new setting (probably as progressive epona song)

leggettc18 committed:
Fix typo

leggettc18 committed:
Remove commented code

leggettc18 committed:
Improve toggle functionality

leggettc18 committed:
Re-introduce cosmetic matching for ocarina buttons

leggettc18 commented:
> Not for this PR but I'm still trying to figure out if there's a good way to visually indicate that this option exists (unless I missed it here), curious if you had any ideas

I think I could do something similar to the Navi button, except it says "Misc." or maybe "Tracker" or "Next" and then when on this page it could say "Quest" or "Prev", but I would need some custom assets for that. I was going to get the MVP in so it was easier for more people to mess around with and then get some of that going.

leggettc18 committed:
Revert some unnecessary formatting changes

leggettc18 committed:
Fix cursor/page turning issue

More improvements to come here (drawing arrows, custom text at the bottom, etc.)

leggettc18 committed:
Fix some more formatting changes

leggettc18 committed:
One last batch of formatting reverts
